src/tools/crawling_tools.py

```
@@ -384,6 +475,252 @@ async def search_code_examples(ctx: Context, query: str, source_id: str = None,
+        logger.warning(f"🛑🛑🛑 SEARCH_CODE_EXAMPLES CANCELLED for: {query} 🛑🛑🛑")
+        logger.warning("🔍 Current stack trace on cancellation:")
+        for line in traceback.format_stack():
+            logger.warning(f"  {line.strip()}")
+        logger.warning("🛑 Re-raising CancelledError to MCP framework")
+        raise
     except Exception as e:
+        await ctx.error(f"Code search failed for '{query}': {str(e)}")
         return json.dumps({"success": False, "query": query, "error": str(e)}, indent=2)
 
+@mcp.tool()
+async def crawl_dir(
```

Ruff (0.12.2)

`crawl_dir` is too complex (30 > 10)

(C901)